pkg/handler3/handler.go

```diff
@@ -92,94 +134,91 @@ func constructServerRelativeURL(gvString, etag string) string {
+			}
 		}
-		discovery.Paths[gvString] = OpenAPIV3DiscoveryGroupVersion{
-			ServerRelativeURL: constructServerRelativeURL(gvString, string(etagBytes)),
+		j, err := json.Marshal(discovery)
+		if err != nil {
+			return cached.NewResultErr[[]byte](err)
 		}
-	}
-	j, err := json.Marshal(discovery)
-	if err != nil {
-		return nil, err
-	}
-	return j, nil
+		return cached.NewResultOK(j, computeETag(j))
+	}, caches)
 }
 
 func (o *OpenAPIService) getSingleGroupBytes(getType string, group string) ([]byte, string, time.Time, error) {
-	o.rwMutex.RLock()
-	defer o.rwMutex.RUnlock()
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
 	v, ok := o.v3Schema[group]
 	if !ok {
 		return nil, "", time.Now(), fmt.Errorf("Cannot find CRD group %s", group)
 	}
 	if getType == subTypeJSON {
-		specBytes, err := v.jsonCache.Get()
-		if err != nil {
-			return nil, "", v.lastModified, err
-		}
-		etagBytes, err := v.etagCache.Get()
-		return specBytes, string(etagBytes), v.lastModified, err
+		result := v.jsonCache.Get()
+		return result.Data, result.Etag, v.lastModified, result.Err
 	} else if getType == subTypeProtobuf {
-		specPb, err := v.pbCache.Get()
-		if err != nil {
-			return nil, "", v.lastModified, err
-		}
-		etagBytes, err := v.etagCache.Get()
-		return specPb, string(etagBytes), v.lastModified, err
+		result := v.pbCache.Get()
+		return result.Data, result.Etag, v.lastModified, result.Err
 	}
 	return nil, "", time.Now(), fmt.Errorf("Invalid accept clause %s", getType)
 }
 
-func (o *OpenAPIService) UpdateGroupVersion(group string, openapi *spec3.OpenAPI) (err error) {
-	o.rwMutex.Lock()
-	defer o.rwMutex.Unlock()
-
+// UpdateGroupVersionLazy adds or updates an existing group with the new cached.
+func (o *OpenAPIService) UpdateGroupVersionLazy(group string, openapi cached.Data[*spec3.OpenAPI]) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
 	if _, ok := o.v3Schema[group]; !ok {
-		o.v3Schema[group] = &OpenAPIV3Group{}
+		o.v3Schema[group] = newOpenAPIV3Group()
+		// Since there is a new item, we need to re-build the cache map.
+		o.discoveryCache.Replace(o.buildDiscoveryCacheLocked())
 	}
-	return o.v3Schema[group].UpdateSpec(openapi)
+	o.v3Schema[group].UpdateSpec(openapi)
 }
 
-func (o *OpenAPIService) DeleteGroupVersion(group string) {
-	o.rwMutex.Lock()
-	defer o.rwMutex.Unlock()
-	delete(o.v3Schema, group)
+func (o *OpenAPIService) UpdateGroupVersion(group string, openapi *spec3.OpenAPI) {
+	o.UpdateGroupVersionLazy(group, cached.NewResultOK(openapi, uuid.New().String()))
 }
 
-func ToV3ProtoBinary(json []byte) ([]byte, error) {
-	document, err := openapi_v3.ParseDocument(json)
-	if err != nil {
-		return nil, err
-	}
-	return proto.Marshal(document)
+func (o *OpenAPIService) DeleteGroupVersion(group string) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	delete(o.v3Schema, group)
+	// Rebuild the merge cache map since the items have changed.
+	o.discoveryCache.Replace(o.buildDiscoveryCacheLocked())
```

Comment:
I feel like there is a missed opportunity here to prepend the `discoveryCache` pipeline with the v3Schema somehow to avoid having to remember to update discoveryCache (like etag for v3Schema changes as you add/remove/update), but I haven't thought it though completely.

Reply:
I think I see what you mean, let me try to consider it

Reply:
Yeah, I think we need the lock on `v3Schema` unfortunately (and I can't think of a way around it), and we need to invalidate some cache anytime the dict changes anyway. Though to be fair, I think this is still absolutely great:
1. If you change the dict itself, it'll regenerate the top-level document, but will not recompute anything for sub-documents that haven't changed
2. If any of the sub-document changes, this will invalidate the top-level document, and all other sub-documents are not recomputed.
Basically, only the things that need to be re-computed will ever be, it's as good as it gets.